### PR TITLE
Port modeling layers

### PR DESCRIPTION
**Pull request by mattdangerw**
🚧 This is an experimental feature branch, more details soon. 🚧

### CODE REVIEW

File: keras_nlp/backend/ops.py [hunk 1]
```diff
@@ -22,3 +25,26 @@
     from keras_core.src.backend.tensorflow.math import *  # noqa: F403, F401
     from keras_core.src.backend.tensorflow.nn import *  # noqa: F403, F401
     from keras_core.src.backend.tensorflow.numpy import *  # noqa: F403, F401
+
+
+if keras_core.config.backend() == "tensorflow" or not multi_backend():
+
+    def take_along_axis(x, indices, axis=None):
```
Comment (mattdangerw) [inline keras_nlp/backend/ops.py:32]: This is the one nasty workaround for the `ops` layer I have left. I don't really like it yet, so haven't moved it into core, but we should find a solution for a `gather` like op in keras-core that is more friendly to dynamic shapes in tf.

Without this, shapes are popping out of `take_along_axis` with fully unknown shapes (e.g. `None, None, None`), which is breaking other parts of `fit()` on our models.

File: keras_nlp/conftest.py [hunk 2]
```diff
@@ -95,6 +95,10 @@ def pytest_collection_modifyitems(config, items):
     run_tpu = config.getoption("--run_tpu")
 
     # Messages to annotate skipped tests with.
+    skip_tf_saving_test = pytest.mark.skipif(
```
Comment (jbischof) [inline keras_nlp/conftest.py:98]: @ianstenbit removed all the tf saving tests in his PR

File: keras_nlp/layers/modeling/cached_multi_head_attention.py [hunk 3]
```diff
@@ -87,8 +85,12 @@ def call(
         cache=None,
         cache_update_index=None,
     ):
-        if not self._built_from_signature:
+        if (
+            hasattr(self, "_build_from_signature")
```
Comment (jbischof) [inline keras_nlp/layers/modeling/cached_multi_head_attention.py:89]: Did we change the API?

File: keras_nlp/layers/modeling/position_embedding.py [hunk 4]
```diff
@@ -46,8 +45,7 @@ class PositionEmbedding(keras.layers.Layer):
 
     Called directly on input.
     >>> layer = keras_nlp.layers.PositionEmbedding(sequence_length=10)
-    >>> layer(tf.zeros((8, 10, 16))).shape
-    TensorShape([8, 10, 16])
+    >>> layer(tf.zeros((8, 10, 16)))
```
Comment (jbischof) [inline keras_nlp/layers/modeling/position_embedding.py:48]: `np.zeros`?

File: keras_nlp/layers/modeling/cached_multi_head_attention.py [hunk 5]
```diff
@@ -101,17 +103,18 @@ def call(
         # cache at the specified index. `cache = None` handles the training
         # case, where we don't use the cache at all.
         if cache is not None:
-            key_cache, value_cache = tf.unstack(cache, axis=1)
+            key_cache = cache[:, 0, ...]
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/cached_multi_head_attention.py:106]: (optional): ops.split might be a bit nicer here

File: keras_nlp/layers/modeling/f_net_encoder.py [hunk 6]
```diff
@@ -93,17 +96,19 @@ def __init__(
         self.kernel_initializer = keras.initializers.get(kernel_initializer)
         self.bias_initializer = keras.initializers.get(bias_initializer)
 
-    def build(self, input_shape):
+    def build(self, inputs_shape):
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/f_net_encoder.py:99]: Why the rename here?

**Review by jbischof — APPROVED** [review]
Some nice cleanups in there

File: keras_nlp/layers/modeling/f_net_encoder_test.py [hunk 7]
```diff
@@ -78,45 +80,14 @@ def test_one_training_step_of_f_net_encoder(self):
...
+        loss = keras.losses.BinaryCrossentropy(from_logits=False)
         optimizer = keras.optimizers.Adam()
-        with tf.GradientTape() as tape:
-            pred = model(data)
-            loss = loss_fn(label, pred)
-        grad = tape.gradient(loss, model.trainable_variables)
-        self.assertGreater(len(grad), 1)
-        optimizer.apply_gradients(zip(grad, model.trainable_variables))
-
-    def test_checkpointing_f_net_encoder(self):
```
Comment (ianstenbit) [inline keras_nlp/layers/modeling/f_net_encoder_test.py:93]: Is this still useful as a `tf_only` test?

**Review by ianstenbit — APPROVED** [review]
🎊

File: keras_nlp/layers/modeling/f_net_encoder.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/f_net_encoder.py:99]: no longer strictly necessary, at one point this was with keras-core.

but the general rule is that build arguments should match call arguments + `_shape`. in the one argument case, any name fine.

File: keras_nlp/layers/modeling/f_net_encoder_test.py (hunk 7 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/f_net_encoder_test.py:93]: IDK, are you testing any checkpoint stuff in KerasCV? This is kind of just a weird holdover we don't do consistently.

We have so many serialization things we could check, I just thought this was a little much. Better to focus on Keras saving/loading.

But we definitely could keep testing checkpoints if we want?

File: keras_nlp/layers/modeling/cached_multi_head_attention.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/cached_multi_head_attention.py:89]: Short version, this is a gross bit of `tf.keras` we don't need to deal with in `keras-core`.

Long version, `tf.keras` has issues supporting multiple arguments to call/build, resulting in the `MultiHeadAttention` layer working around the normal `build` function with `_build_from_signature` and hitting some issues with serialization. This was not a hack that got propogated to keras-core thankfully, normal `build` can now support multiple build arguments.

Can post a more in the weeds explainer if you would like

File: keras_nlp/layers/modeling/cached_multi_head_attention.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/cached_multi_head_attention.py:106]: Was not able to get this working. `split(cache, 2, axis=1)` is not specific enough shape wise. Is that the right invocation you were thinking of?

File: keras_nlp/layers/modeling/cached_multi_head_attention.py (hunk 3 above)
Comment (jbischof) [reply keras_nlp/layers/modeling/cached_multi_head_attention.py:89]: Yeah I've seen some bugs around that one. Good fix

File: keras_nlp/layers/modeling/f_net_encoder_test.py (hunk 7 above)
Comment (mattdangerw) [reply keras_nlp/layers/modeling/f_net_encoder_test.py:93]: talked offline. let's ditch these

File: keras_nlp/layers/modeling/cached_multi_head_attention.py (hunk 5 above)
Comment (ianstenbit) [reply keras_nlp/layers/modeling/cached_multi_head_attention.py:106]: Ahh yes, that probably won't work then. `split(cache, [1], axis=1)` would probably give you a defined shape for `key_cache` but not `value_cache`. So this seems good as-is

File: keras_nlp/conftest.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/conftest.py]: did this too!

**Review by fchollet — APPROVED** [review]
LGTM
